Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
-        auto* element = dynamicDowncast<Element>(*node);
-        if (node->hasTagName(bdiTag) || node->hasTagName(scriptTag) || node->hasTagName(styleTag)
-            || (element && element->isTextField())) {
-            node = NodeTraversal::nextSkippingChildren(*node, this);
+std::optional<TextDirection> Element::computeAutoTextDirection() const
+{
+    if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(*this))
+        return computeTextFormControlElementTextDirection(*textFormControl);
+
+    if (auto* slotElement = dynamicDowncast<HTMLSlotElement>(*this))
+        return computeSlotElementTextDirection(*slotElement);
+
+    return computeElementContainedTextAutoDirection(*this);
+}
+
+std::optional<TextDirection> Element::computeTextDirection() const
+{
+    if (!(selfOrPrecedingNodesAffectDirAuto() && hasAutoTextDirectionState()))
+        return std::nullopt;
+    return computeElementTextDirection(*this, elementTextDirectionState(*this));
+}
+
+void Element::updateEffectiveTextDirectionOfDescendants(std::optional<TextDirection> direction, Element* exclude)
+{
+    for (auto it = descendantsOfType<Element>(*this).begin(); it; ) {
+        Ref child = *it;
+
+        if (elementHasValidTextDirectionState(child)) {
+            it.traverseNextSkippingChildren();
             continue;
         }
 
-        // Skip elements with valid dir attribute
-        if (element) {
-            if (isValidDirValue(element->attributeWithoutSynchronization(dirAttr))) {
-                node = NodeTraversal::nextSkippingChildren(*element, this);
-                continue;
-            }
-        }
+        if (child.ptr() != exclude)
+            updateElementAndShadowEffectiveTextDirection(child, direction);
 
-        if (node->isTextNode()) {
-            if (auto direction = node->textContent(true).defaultWritingDirection())
-                return { *direction == U_LEFT_TO_RIGHT ? TextDirection::LTR : TextDirection::RTL, node };
-        }
-        node = NodeTraversal::next(*node, this);
+        it.traverseNext();
     }
-    return { TextDirection::LTR, nullptr };
 }
 
-void Element::adjustDirectionalityIfNeededAfterChildAttributeChanged(Element*)
+void Element::updateEffectiveTextDirectionOfAncestor(Element* exclude)
 {
     ASSERT(selfOrPrecedingNodesAffectDirAuto());
-    for (Ref element : lineageOfType<Element>(*this)) {
-        if (elementAffectsDirectionality(element)) {
-            ASSERT(element->hasDirectionAuto());
-            element->updateEffectiveDirectionalityOfDirAuto();
-            break;
-        }
+    for (Ref ancestor : lineageOfType<Element>(*this)) {
+        if (!elementHasAutoTextDirectionState(ancestor))
+            continue;
+        updateElementTextDirection(ancestor, TextDirectionState::Auto, exclude);
+        break;
     }
 }
 
+void Element::updateTextDirection(std::optional<AtomString> newValue)
+{
+    auto value = newValue.value_or(attributeWithoutSynchronization(dirAttr));
```

Comment:
This is a very confusing function interface. I think it's better if the caller always passed in newValue instead of conditionally getting the attribute value like this inside the function.

Reply:
Done.

I put back the original function `Element::dirAttributeChanged()`. So I was able to remove the `newValue` argument form this function.